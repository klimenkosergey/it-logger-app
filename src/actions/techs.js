export const getTechs = () => async dispatch => {
  try {
    const res = await fetch('/techs');
    const data = await res.json();

    dispatch({ type: 'GET_TECHS', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error getting technicians' });
  }
};

export const addTech = tech => async dispatch => {
  try {
    const res = await fetch('/techs', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(tech)
    });
    const data = await res.json();

    dispatch({ type: 'ADD_TECH', payload: data });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error adding technician' });
  }
};

export const deleteTech = id => async dispatch => {
  try {
    await fetch(`/techs/${id}`, {
      method: 'DELETE'
    });

    dispatch({ type: 'DELETE_TECH', payload: id });
  } catch (error) {
    dispatch({ type: 'SET_ERROR', payload: 'Error deleting technician' });
  }
};
