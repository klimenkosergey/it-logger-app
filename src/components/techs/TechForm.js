import React, { Fragment, useState } from 'react';

const TechForm = props => {
  const [firstName, setFirstName] = useState(props.firstName || '');
  const [lastName, setLastName] = useState(props.lastName || '');

  const onSubmit = e => {
    e.preventDefault();
    props.onSubmit(firstName, lastName);

    // Clear inputs
    setFirstName('');
    setLastName('');
  };

  return (
    <Fragment>
      <div className='modal-content'>
        <h4 className='center'>
          {!props.firstName && !props.lastName ? 'Add' : 'Edit'} Technician
        </h4>
        <div className='col s12'>
          <div className='row'>
            <div className='input-field col s6'>
              <input
                type='text'
                name='firstName'
                value={firstName}
                onChange={e => setFirstName(e.target.value)}
              />
              <label htmlFor='firstName' className='active'>
                First Name
              </label>
            </div>
            <div className='input-field col s6'>
              <input
                type='text'
                name='lastName'
                value={lastName}
                onChange={e => setLastName(e.target.value)}
              />
              <label htmlFor='lastName' className='active'>
                Last Name
              </label>
            </div>
          </div>
        </div>
      </div>
      <div className='modal-footer'>
        <a href='#!' onClick={onSubmit} className='waves-light blue btn'>
          Submit
        </a>
      </div>
    </Fragment>
  );
};

export default TechForm;
