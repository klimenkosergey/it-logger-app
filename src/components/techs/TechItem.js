import React from 'react';
import PropTypes from 'prop-types';

const TechItem = ({ tech, onEdit, onDelete }) => (
  <li className='collection-item'>
    <a href='#!' onClick={onEdit} className='blue-text'>
      {tech.firstName} {tech.lastName}
    </a>
    <span className='secondary-content'>
      <a
        href='#!'
        className='tech-delete'
        onClick={e => {
          e.preventDefault();
          onDelete(tech.id, tech.firstName, tech.lastName);
        }}
      >
        <i className='material-icons grey-text'>delete</i>
      </a>
    </span>
  </li>
);

TechItem.propTypes = {
  tech: PropTypes.object.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default TechItem;
