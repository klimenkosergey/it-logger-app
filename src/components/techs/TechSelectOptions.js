import React, { Fragment } from 'react';
import { connect } from 'react-redux';

const TechSelectOptions = ({ techs }) => {
  return (
    <Fragment>
      {techs.length > 0 &&
        techs.map(tech => (
          <option
            key={tech.id}
            value={`${tech.firstName} ${tech.lastName}`}
          >{`${tech.firstName} ${tech.lastName}`}</option>
        ))}
    </Fragment>
  );
};

const mapStateToProps = state => ({
  techs: state.techs.techs
});

const ConnectedTechSelectOptions = connect(mapStateToProps)(TechSelectOptions);

export { TechSelectOptions, ConnectedTechSelectOptions as default };
