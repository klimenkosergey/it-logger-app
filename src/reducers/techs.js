const techsReducerDefaultState = {
  techs: [],
  loading: true,
  error: null
};

const techsReducer = (state = techsReducerDefaultState, action) => {
  switch (action.type) {
    case 'GET_TECHS':
      return {
        ...state,
        loading: false,
        techs: action.payload
      };
    case 'ADD_TECH':
      return {
        ...state,
        techs: [...state.techs, action.payload]
      };
    case 'EDIT_TECH':
      return {
        ...state,
        techs: state.techs.map(tech =>
          tech.id === action.payload.id ? action.payload : tech
        )
      };
    case 'DELETE_TECH':
      return {
        ...state,
        techs: state.techs.filter(tech => tech.id !== action.payload)
      };
    case 'SET_ERROR':
      console.error(action.payload);
      return {
        ...state,
        error: action.payload
      };
    default:
      return state;
  }
};

export default techsReducer;
