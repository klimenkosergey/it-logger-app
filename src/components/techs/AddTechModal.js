import React from 'react';
import { connect } from 'react-redux';
import M from 'materialize-css/dist/js/materialize';

import { addTech } from '../../actions/techs';
import TechForm from './TechForm';

const AddTechModal = ({ addTech }) => {
  const onSubmit = (firstName, lastName) => {
    const modalElem = document.querySelector('#add-tech-modal');
    const instance = M.Modal.getInstance(modalElem);

    if (firstName === '' || lastName === '') {
      M.toast({ html: 'Please enter first and last names' });
    } else {
      instance.close();
      M.toast({ html: `Added ${firstName} ${lastName} as technician` });
      addTech({ firstName, lastName });
    }
  };

  return (
    <div id='add-tech-modal' className='modal'>
      <TechForm onSubmit={onSubmit} />
    </div>
  );
};

const ConnectedAddTechModal = connect(
  null,
  { addTech }
)(AddTechModal);

export { AddTechModal, ConnectedAddTechModal as default };
