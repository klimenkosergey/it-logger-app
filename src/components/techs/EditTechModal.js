import React from 'react';
import M from 'materialize-css/dist/js/materialize';

import TechForm from './TechForm';

const EditTechModal = () => {
  const onSubmit = (firstName, lastName) => {
    const modalElem = document.querySelector('#edit-tech-modal');
    const instance = M.Modal.getInstance(modalElem);

    if (firstName === '' || lastName === '') {
      M.toast({ html: 'Please enter first and last names' });
    } else {
      instance.close();
    }
  };

  return (
    <div id='edit-tech-modal' className='modal'>
      <TechForm onSubmit={onSubmit} />
    </div>
  );
};

export default EditTechModal;
