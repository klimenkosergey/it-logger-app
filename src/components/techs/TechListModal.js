import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import M from 'materialize-css/dist/js/materialize';

import { getTechs, deleteTech } from '../../actions/techs';
import TechItem from './TechItem';

const TechListModal = ({ techs, loading, getTechs, deleteTech }) => {
  useEffect(() => {
    getTechs();
    // eslint-disable-next-line
  }, []);

  const onEdit = e => {
    e.preventDefault();
    M.Modal.getInstance(document.querySelector('#tech-list-modal')).close();
    M.Modal.getInstance(document.querySelector('#edit-tech-modal')).open();
  };

  const onDelete = (id, firstName, lastName) => {
    M.toast({ html: `Removed technician ${firstName} ${lastName}` });
    deleteTech(id);
  };

  return (
    <div id='tech-list-modal' className='modal'>
      <div className='modal-content'>
        <h4 className='center'>Technician List</h4>
        {loading ? (
          <p className='center'>Loading...</p>
        ) : (
          <ul className='collection'>
            {techs.length === 0 ? (
              <p className='center'>No technicians to show...</p>
            ) : (
              techs.map(tech => (
                <TechItem
                  key={tech.id}
                  tech={tech}
                  onEdit={onEdit}
                  onDelete={onDelete}
                />
              ))
            )}
          </ul>
        )}
      </div>
      <div className='modal-footer'>
        <a href='#!' className='modal-close waves-light blue btn'>
          Close
        </a>
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  techs: state.techs.techs,
  loading: state.techs.loading
});

const ConnectedTechListModal = connect(
  mapStateToProps,
  { getTechs, deleteTech }
)(TechListModal);

export { TechListModal, ConnectedTechListModal as default };
