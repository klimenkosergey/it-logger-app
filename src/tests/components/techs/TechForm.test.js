import React from 'react';
import { shallow } from 'enzyme';

import techs from '../../fixtures/techs';
import TechForm from '../../../components/techs/TechForm';

let wrapper, onSubmitSpy;

beforeEach(() => {
  onSubmitSpy = jest.fn();
  wrapper = shallow(<TechForm onSubmit={onSubmitSpy} />);
});

it('should render TechForm correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should render TechForm with provided tech', () => {
  wrapper = shallow(<TechForm onSubmit={onSubmitSpy} {...techs[0]} />);
  expect(wrapper).toMatchSnapshot();
});

it('should handle onSubmit', () => {
  wrapper.find('.btn').simulate('click', { preventDefault: jest.fn() });
  expect(onSubmitSpy).toHaveBeenCalled();
});

it('should change firstName on input change', () => {
  const firstName = 'Bruce';
  wrapper
    .find('input[name="firstName"]')
    .simulate('change', { target: { value: firstName } });
  expect(wrapper.find('input[name="firstName"]').prop('value')).toBe(firstName);
});

it('should change lastName on input change', () => {
  const lastName = 'Banner';
  wrapper
    .find('input[name="lastName"]')
    .simulate('change', { target: { value: lastName } });
  expect(wrapper.find('input[name="lastName"]').prop('value')).toBe(lastName);
});
