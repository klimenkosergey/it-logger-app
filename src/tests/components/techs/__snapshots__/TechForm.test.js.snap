// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TechForm correctly 1`] = `
<Fragment>
  <div
    className="modal-content"
  >
    <h4
      className="center"
    >
      Add
       Technician
    </h4>
    <div
      className="col s12"
    >
      <div
        className="row"
      >
        <div
          className="input-field col s6"
        >
          <input
            name="firstName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="active"
            htmlFor="firstName"
          >
            First Name
          </label>
        </div>
        <div
          className="input-field col s6"
        >
          <input
            name="lastName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="active"
            htmlFor="lastName"
          >
            Last Name
          </label>
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal-footer"
  >
    <a
      className="waves-light blue btn"
      href="#!"
      onClick={[Function]}
    >
      Submit
    </a>
  </div>
</Fragment>
`;

exports[`should render TechForm with provided tech 1`] = `
<Fragment>
  <div
    className="modal-content"
  >
    <h4
      className="center"
    >
      Edit
       Technician
    </h4>
    <div
      className="col s12"
    >
      <div
        className="row"
      >
        <div
          className="input-field col s6"
        >
          <input
            name="firstName"
            onChange={[Function]}
            type="text"
            value="Tony"
          />
          <label
            className="active"
            htmlFor="firstName"
          >
            First Name
          </label>
        </div>
        <div
          className="input-field col s6"
        >
          <input
            name="lastName"
            onChange={[Function]}
            type="text"
            value="Stark"
          />
          <label
            className="active"
            htmlFor="lastName"
          >
            Last Name
          </label>
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal-footer"
  >
    <a
      className="waves-light blue btn"
      href="#!"
      onClick={[Function]}
    >
      Submit
    </a>
  </div>
</Fragment>
`;
